### PR TITLE
Botan improvements and misc fixes

### PR DESCRIPTION
**Pull request by jacklul**
Botan.io:
- moved to Guzzle
- user can set own timeout for Botan requests `$telegram->enableBotan('TOKEN', ['timeout' => 5])`, defaults to 3
- added shortener usage example
- stated potential usage for `public static $command` in docblock
- fixed botan not working with getUpdates

Other:
- Fixed `Call to a member function getBody() on null` happening on empty or no replies (currently visible on Windows, will be replaced with 'invalid response' exception which is better)
- Fixed code with PHP Code Beautifier and Fixer
- Added example usage of editMessageText 
(I think it would be a good idea to make more usage/command examples based on past closed issues)

### CODE REVIEW

File: src/Commands/AdminCommands/SendtochannelCommand.php [hunk 1]
```diff
@@ -81,7 +81,7 @@ public function execute()
         $notes = &$this->conversation->notes;
         !is_array($notes) && $notes = [];
 
-        $channels = (array)$this->getConfig('your_channel');
+        $channels = (array) $this->getConfig('your_channel');
```
Comment (noplanman) [inline src/Commands/AdminCommands/SendtochannelCommand.php]: We should decide on a specific format and keep it that way, otherwise these little things will be overwritten all the time by each other.
At the moment there is some mix of this, some places are one way, others another.
I'll set up an issue for us to discuss this briefly, as it's not such a big deal...

File: src/Botan.php [hunk 2]
```diff
@@ -181,33 +227,37 @@ public static function shortenUrl($url, $user_id)
...
+                )
+            );
 
-        $context  = stream_context_create($options);
-        $response = @file_get_contents($request, false, $context);
+            $response = (string) $response->getBody();
+        } catch (RequestException $e) {
+            $response = ($e->getResponse()) ? (string) $e->getResponse()->getBody() : '';
+        } finally {
+            if (!filter_var($response, FILTER_VALIDATE_URL) === false) {
```
Comment (noplanman) [inline src/Botan.php]: Just using `filter_var($response, FILTER_VALIDATE_URL) !== false` is easier to read.

File: src/Botan.php [hunk 3]
```diff
@@ -181,33 +227,37 @@ public static function shortenUrl($url, $user_id)
...
-        if (!empty($cached[0]['short_url'])) {
-            return $cached[0]['short_url'];
-        }
 
-        $request = str_replace(
-            ['#TOKEN', '#UID', '#URL'],
-            [self::$token, $user_id, urlencode($url)],
-            self::$shortener_url
-        );
+        if (!empty($cached)) {
```
Comment (noplanman) [inline src/Botan.php]: `!empty()` is redundant here, as `$cached` gets defined above.
You could also just do:
```php
if ($cached = BotanDB::selectShortUrl($user_id, $url)) {
    return $cached;
}
```

File: src/Botan.php [hunk 4]
```diff
@@ -20,82 +23,107 @@
...
         self::$command = '';
 
-        $obj  = json_decode($input, true);
+        // For now, this is the only way
+        $update_data = (array) $update;
+
         $data = [];
-        if (isset($obj['message'])) {
-            $data       = $obj['message'];
+        if ($update->getMessage()) {
```
Comment (noplanman) [inline src/Botan.php]: Would be nice if we could find a way to shorten this epic `if else` block and make it more dynamic.
Not priority though, more a though since you've touched code here 😃

File: src/Telegram.php [hunk 5]
```diff
@@ -815,14 +816,15 @@ protected function ucfirstUnicode($str, $encoding = 'UTF-8')
...
      *
-     * @param  $token
+     * @param  string $token
+     * @param  integer $timeout
      *
      * @return \Longman\TelegramBot\Telegram
      * @throws \Longman\TelegramBot\Exception\TelegramException
      */
-    public function enableBotan($token)
+    public function enableBotan($token, $timeout = 3)
```
Comment (noplanman) [inline src/Telegram.php]: I'm just wondering if we might want to have an array here for extra parameters, to prevent issues in the future if new parameters might become available.
Having an array, we just add indexes and previous code still works. Adding extra parameters to the method becomes difficult to manage.

Here an example of the problem I'm trying to explain:
https://github.com/akalongman/php-telegram-bot/pull/346/files#diff-04c6e90faac2675aa89e2176d2eec7d8L245

**Comment by jacklul** [comment]
Which line?

File: src/Botan.php [hunk 6]
```diff
@@ -20,152 +23,195 @@
...
+            'message' => 'Message',
+            'edited_message' => 'Edited Message',
+            'channel_post' => 'Channel Post',
+            'edited_channel_post' => 'Edited Channel Post',
+            'inline_query' => 'Inline Query',
+            'chosen_inline_result' => 'Chosen Inline Result',
+            'callback_query' => 'Callback Query'
+        ];
+
+        if (in_array($update_type, ['message', 'edited_message', 'channel_post', 'edited_channel_post', 'inline_query', 'chosen_inline_result', 'callback_query'], true)) {
```
Comment (noplanman) [inline src/Botan.php]: You could use this instead:
```php
if (array_key_exists($update_type, $update_object_names)) {...
```

Right?

Comment (jacklul) [inline src/Botan.php]: Oh, that's what you meant!

File: examples/getUpdatesCLI.php [hunk 7]
```diff
@@ -62,7 +62,9 @@
     //$telegram->setUploadPath('../Upload');
 
     // Botan.io integration
+    // Second argument is optional maximum timeout
```
Comment (noplanman) [inline examples/getUpdatesCLI.php]: Nitty-gritty: `Second argument are options`

*(same for hook.php)*

File: src/Botan.php [hunk 8]
```diff
@@ -20,152 +23,195 @@
...
+                    'json' => $data
+                ]
+            );
 
-        if ($responseData['status'] !== 'accepted') {
-            TelegramLog::debug('Botan.io API replied with error: ' . $response);
-        }
+            $response = (string) $response->getBody();
+        } catch (RequestException $e) {
+            $response = ($e->getResponse()) ? (string) $e->getResponse()->getBody() : '';
```
Comment (noplanman) [inline src/Botan.php]: Looking at the Guzzle `RequestException` class, it seems the response always gets set.

Have you run into a situation where this wasn't the case?

Also, I think variable-name-wise, it would be better not to overwrite `$response` with the response body, as that would actually be `$response_body`.
Maybe call and check the `->getBody()` in the `finally` clause? Also to remove the duplication above.

*(same applies to the other similar code further down)*

Comment (jacklul) [inline src/Botan.php]: Yes, currently on Windows there is `can't call getBody() on null` error here and in `Request::execute()`.

File: src/Botan.php [hunk 9]
```diff
@@ -187,17 +187,17 @@ public static function track(Update $update, $command = '')
...
-            $response = ($e->getResponse()) ? (string) $e->getResponse()->getBody() : '';
+            $result = $e->getMessage();
         } finally {
-            $responseData = json_decode($response, true);
+            $responseData = json_decode($result, true);
 
             if ($responseData['status'] !== 'accepted') {
-                if (!empty($response)) {
-                    TelegramLog::debug("Botan.io stats report failed, API reply: $response");
+                if (!empty($result)) {
```
Comment (noplanman) [inline src/Botan.php]: Well, if you're doing it like this, might as well shorten it all the way 😃
```php
TelegramLog::debug('Botan.io stats report failed: ' . ($result ?: 'empty response') . "\n\n");
```

*(same for the bottom one)*

File: src/Botan.php (hunk 8 above)
Comment (noplanman) [inline src/Botan.php]: Oh right, ok... The way you've changed it in the last commit works for Windows nicely now too then?

Comment (jacklul) [inline src/Botan.php]: At least not throwing that error and stopping execution like it did before...

File: src/BotanDB.php [hunk 10]
```diff
@@ -32,8 +32,8 @@ public static function initializeBotanDb()
     /**
```
Comment (noplanman) [inline src/BotanDB.php]: I think it would make sense to use the same parameter order like for `shortenUrl($url, $user_id)`:
so `selectShortUrl($url, $user_id)`

File: src/Botan.php [hunk 11]
```diff
@@ -20,152 +23,191 @@
...
-        $options = [
-            'http' => [
-                'header'        => 'Content-Type: application/json',
-                'method'        => 'POST',
-                'content'       => json_encode($data),
-                'ignore_errors' => true,
-            ],
-        ];
+        // In case there is no from field assign id = 0
+        if (isset($data['from']['id'])) {
```
Comment (noplanman) [inline src/Botan.php]: Maybe shorten this to 1 line too? Up to you.

File: src/Botan.php [hunk 12]
```diff
@@ -20,152 +23,191 @@
...
+            $uid = $data['from']['id'];
+        } else {
+            $uid = 0;
+        }
 
-        $context      = stream_context_create($options);
-        $response     = @file_get_contents($request, false, $context);
-        $responseData = json_decode($response, true);
+        try {
+            $response = self::$client->post(
```
Comment (noplanman) [inline src/Botan.php]: This whole block looks really nice and neat now, well done :1st_place_medal:

File: src/BotanDB.php (hunk 10 above)
Comment (jacklul) [inline src/BotanDB.php]: That was on my mind but it seems to forgot about it, will do!
